### DISCUSSION TITLE
get value without using id

### DISCUSSION THREAD

**Discussion by GerryTo** [appium/appium#20402] [Q&A]
i have some problem when getting value using appium, can we get value without using id ?

**Comment by portyu9** [comment]
Hi @GerryTo 

Yes. You do not need an `id` to get an element's value.

Locate the element using another stable attribute, preferably accessibility ID:

```java
WebElement element =
    driver.findElement(AppiumBy.accessibilityId("username"));

String value = element.getAttribute("value");
```

On Android you can also use UIAutomator:

```java
WebElement element = driver.findElement(
    AppiumBy.androidUIAutomator(
        "new UiSelector().text(\"Username\")"
    )
);
```

or XPath as a fallback:

```java
WebElement element =
    driver.findElement(By.xpath("//android.widget.EditText"));
```

Then use:

```java
element.getText();
element.getAttribute("value");
```

depending on what attribute your platform exposes.

If you control the application, the best long-term fix is still to add a stable accessibility/test identifier rather than depending on XPath or visible text.

Please mark this answer as accepted if it helped, thank you.
